### PR TITLE
Translation

### PR DESCRIPTION
**Pull request by Adimgar**
This is my first attempt to enable support Intllib library from Kaeza, all mods in minetest_game have now Intllib support and includes spanish translations for its texts

### CODE REVIEW

File: mods/beds/functions.lua [hunk 1]
```diff
@@ -34,7 +36,7 @@ local function check_in_beds(players)
 		end
 	end
 
-	return #players > 0
+	return true
```
Comment (est31) [inline mods/beds/functions.lua]: You should change this back. If there is a merge conflict, don't override, but merge it manually.

File: mods/beds/functions.lua [hunk 2]
```diff
@@ -97,13 +99,13 @@ local function update_formspecs(finished)
 
 	if finished then
 		form_n = beds.formspec ..
-			"label[2.7,11; Good morning.]"
+			S("label[2.7,11; Good morning.]")
```
Comment (est31) [inline mods/beds/functions.lua]: this should always be done via code like `"label[2.7,11; " .. S("Good morning.") .. "]"` to make changes of the label properties easy.

**Comment by est31** [comment]
You should also include intllib itself, by adding it as subdirectory inside `mods`.

File: mods/beds/functions.lua [hunk 3]
```diff
@@ -97,13 +99,13 @@ local function update_formspecs(finished)
...
 	else
 		form_n = beds.formspec ..
-			"label[2.2,11;"..tostring(player_in_bed).." of "..tostring(ges).." players are in bed]"	
+			"label[2.2,11;"..tostring(player_in_bed)..S(" of ")..tostring(ges)..S(" players are in bed]")
```
Comment (est31) [inline mods/beds/functions.lua]: This should be better implemented via passing multiple parameters to S, like `S("@1 of @2 players are in bed", tostring(player_in_bed), tostring(ges))`. There is also a way where there is no `tostring`, but I can't remember how, and intllib is badly documented in this regard. I think @ShadowNinja knows how.

**Comment by LeMagnesium** [comment]
This PR is interesting. I did the same translation process months ago (August 2014 if I remember correctly), but then I abandonned this copy : Intllib has many inconvenients, the most visible being that it translates server-side while parsing the code, and not client-side depending on the player's settings. However, if this gets merged (but I'm not really sure it will), it would be a great feature for non-english servers welcoming people speaking only one common language.

**Comment by Xanthin** [comment]
Yay, I never thought there will be any chance for intllib support of the minetest_game mods by default because of its known limits. Will be great, so I don't need to add any updates of the mt_game mods by hand to keep my translation stuff (April 14, LeMagnesium :P). And of course, for LeMagnesium's statement too.
When it got merged I can provide the german text files.
@Adimgar, can you also add template files, so its easier for others to add different translations?

**Comment by LeMagnesium** [comment]
@Xanthin : I did french translation on my side, it won't be a problem to do it again. Yet, without template files, it will be complicated.
@Adimgar : To create the template file just copy in every `locale/` directory the file `es.txt` as `template.txt` and remove the spanish translation. (Just in case you didn't know how to do it)

**Comment by ExcaliburZero** [comment]
I'd like to see this added in. There really needs to be some way to translate mods in by default. Currently the only way to translate mods is to translate the mod and hope that the user has downloaded the correct translation mod for it to work.

Even though this will only handle serve side translation, I think it would be better than offering no translation at all. Also, I think having a default way to handle mod translations would help to convince more modders to translate or add translation support for their mods. That way more mods would have translations, so if a better system with client side translation is created later on then it would be easier to convert mods to it, as it would likely just be a matter of making some modifications to the translation files, since the translations will have already been made.

**Comment by LeMagnesium** [comment]
@ExcaliburZero : I have doubts about the fact that client side translation would be a good thing. Let's imagine two players with different language settings, one in english, and one in german. If the server has many items, and they try to speak about the same one, they won't want to describe it (too many items would match the description for example), and they would have different descriptions for it.. But knowing that servers' players often speak english, or if they don't, they speak only one common language, make server-side translation simplier in this case.
Anyway we're taling about server-side translation, client-side is still a big thing to code (and might need deep engine changes).

**Comment by ExcaliburZero** [comment]
@LeMagnesium I agree, most of the time with servers there will generally just be one language that is used. Client-side translation wouldn't always be consistent in that while the nodes, items, etc. of the server would be translated into the client's language, other things like signs and chat messages would not be translated.

Client side translation would be useful in a few cases, especially for multi-lingual servers, and thus should probably be added at some point. However, I think serve-side translation should be a bigger priority, as it would be useful in more cases.

I think for now we should add in server-side translation and leave client-side translation for another day, as like you said, that would be a whole other issue.

**Comment by Zeno-** [comment]
There is probably no reason why the server should not be done server-side anyway. If gettext() is enabled there is a function call anyway so performance-wise that's not *really* an issue. *But* how will the server know what language the user (client) wants?

**Comment by est31** [comment]
Clientside translation has the advantage of being easier engine wise. I outline below two approaches, the first being what I think @Zeno-  meant, the second the approach which i favour. Sending supported languages is no problem, with protocol v25 we're getting new init packets anyway. Simply a string containing language codes by order of prefference is easy enough.

#### Option all serverside
The server sends the translated strings to the client.
If we take a description of a node for example, which is stored once at the node's definition. If we do `gettext()` inside the table definition, that function is evaluated exactly once. So, either we translate later on when we send the nodedefs to the client, but this will require to add two things server side:
1. the server has to know which nodes are translateable at all
2. the server has to modify the nodedefs for each client, which is especially complicated as its best exposed via lua too.
Advantage: No protocol change required except language transmission
Disadvantage: Vast changes engine side (and lua api side if we want to use lua based translation code)

#### Option clientside
The server sends basic informations about the strings, the client translates them using translation tables sent by the server.
As before the client sends its language(s) to the server at connect or join. The server then sends a list of translateable strings and their versions in the language of the client. If the client wants to display some string, it checks whether it can be translated.

The major advantage of this is that the server doesn't have to keep language specific data structures for every client except the tables. There is a basic version of this and a less basic version. The basic version is that the client only checks translateable strings on a 1:1 basis with the database. The less basic version is that the strings include a special command character (let's chose a nonprintable ASCII here :) ), a format string, and a list of values. It would look like this: ":earth_africa:@1 of @2 persons sleep in bed:factory:2:factory:3". Of course, legacy clients will display the :earth_africa: part too, so the server would still need to store separate data structures (but only two!) for the non basic version, or use tricks like "2 of 3 persons sleep in bed\0:earth_africa:@1 of @2 persons sleep in bed:factory:2:factory:3". We could first implement the basic version, and then later the non basic version.

**Comment by Adimgar** [comment]
Thank you all for your comments, I'm very pleased that you find this useful and willing to give it a try.
The original idea of this was that people can offer a server in a local language, like if someone opens one in Italy people can see it's content in italian, but now you're going further with client side translations and all that stuff, which would be a very good thing to see in this project. Thnaks again and keep on the great job you're doing.
I'll add templates asap, thanks for pointing it out.
I still have some issues with strings containing spaces, they are not handled corectly by intllib, i was thinking in change the lines to params like in beds (the @1 of @2 persons... way) but since now intllib is included in minetest_game i believe it would be better if it gets improved a little so it can handle all those strings better, like _game having it's own intllib version, but i don't know if i'm able to do so at this moment, i'm barely getting familiar with lua and all this stuff, any advices on this are very welcome, as well as any others.
I'll keep on it and see what happens, thank you all again.

**Comment by Adimgar** [comment]
I've tested/checked all I could imagine and everything looks fine, I believe it's usable now. I'll leave it at this point and wait for any comments, thank you all.

**Comment by ShadowNinja** [comment]
> You should also include intllib itself, by adding it as subdirectory inside mods.

Please no.  This makes it 10x harder to distribute any updates.  And it's made worse by a dozen other games hat copy `minetest_game` near-verbatim.


This is obviously not ideal since the translations are entirely server-side, but it's better than nothing and works perfectly in singleplayer.

Note: I don't have time to look over the code right now, but I approve the concept. :-)

**Comment by est31** [comment]
Why is it harder then to distribute updates?

**Comment by Adimgar** [comment]
Glad o see you around @ShadowNinja 
Including Intllib in _game actually conflicts when the user has the mod in their local mods folder, it should be either removed or renamed (or include it locally in some other way) to avoid dependances error messages I believe. Nevertheless, the dependance on the lib is optional, if the user doesn't have it, it simply won't be translations

**Comment by LeMagnesium** [comment]
> Including Intllib in _game actually conflicts when the user has the mod in their local mods folder, i believe it should be either removed or renamed

In this case, maybe remove it from _game, but let the translations, so they are optional, and if the player wants translation on his/her server, he just has to install intllib. This means we must write somewhere in the game's documentation how to get it translated.

**Comment by est31** [comment]
I'm still waiting for a reason why shipping the mod would be bad. The fact that it might conflict with an already installed version of the mod doesn't count, in my opinion, as this is true for all mods inside _game.

**Comment by sfan5** [comment]
> The fact that it might conflict with an already installed version of the mod doesn't count, in my opinion, as this is true for *all* mods inside _game.

I don't think anyone has `bones`, `bucket`, `creative`, `default`, `doors`, `fire` or  `give_initial_stuff` installed as a mod.

**Comment by Adimgar** [comment]
The only reason I see is to avoid the waring messages and maybe get a lot of people asking about them, but they're just warnings after all. It'll be done as you all like :)

**Comment by Adimgar** [comment]
I'd like to ask for some advice on my last commit, I created individual namespaces for each mod, because some of them (beds and some others) seemed to conflict with the mod's default functions when using the original namespaces for intllib. By doing it this way causes some memory usage issues or any other issue?
Thanks in advance.

**Comment by est31** [comment]
If, as it seems, there are concerns regarding conflicts with existing installs of the `intllib`, we should instead ship with a renamed version, e.g. like `i18nlib`. I mean its much more complicated for the average user when you will first have to install a mod to get singleplayer translated. We shouldn't (soft or not) depend on outside mods more than neccessary.

**Comment by est31** [comment]
About your namespace question @Adimgar, perhaps you want to only keep the `beds.intllib = S` line, and remove the `beds = {}`, as that one is double. Then extra `ibeds` namespaces aren't needed.

**Comment by nerzhul** [comment]
If you want client side translation based on server side a good idea is use some games approach, using a lookup for names, based on IDs (this needs a new packet). you send TOSERVER_LOOKUP_TRANSLATION (std::string country, std::string entity) for example and the servers lookup it's translation tables and return the result.

**Comment by est31** [comment]
Why not TOCLIENT_TRANSLATION_TABLE (std::string table), and including the language inside the init packets? Otherwise the protocol overhead (20 bytes for ipv4, 40 bytes for ipv6) is larger than the actual data, and its not needed in this case.

**Comment by nerzhul** [comment]
It depend of what we want. If we set the language at session init (not in _INIT packet but after _AUTH, it's better) we don't need it in the lookup, right. For the packet you tell, why not, because we are sending every nodedef to client at init it could be very good to send an extra packet to send all translations. It's not the perfect situation because client load everything, but it's already the case, then ok.
TOCLIENT_TRANSLATIONS = (u32 nb + nb * (std::string key + std::string translated_string) is a good structure.

File: mods/beds/init.lua [hunk 4]
```diff
@@ -3,9 +3,18 @@ beds.player = {}
 beds.pos = {}
 beds.spawn = {}
 
+-- Intllib
+local S
+if minetest.get_modpath("intllib") then
+	S = intllib.Getter()
+else
+	S = function(s) return s end
```
Comment (ShadowNinja) [inline mods/beds/init.lua]: If you're using `@` replacements you need to use the longer more recent boilerplate function.

**Comment by ShadowNinja** [comment]
> Why is it harder then to distribute updates?

Because every time me or kaeza makes a change to intllib I'll have to push the change to this repo too.
Bundling it would be O.K. with me iff it was included as a submodule.  That way it isn't duplicated and users can chose if they want it in the game.

File: mods/beds/init.lua [hunk 5]
```diff
@@ -3,9 +3,19 @@ beds.player = {}
 beds.pos = {}
 beds.spawn = {}
 
+-- Intllib
+ibeds = {}
```
Comment (ShadowNinja) [inline mods/beds/init.lua]: Why do you create an `ibeds` table and then only store the translator in it?  You might as well just make a global `beds_gettext`.  Why not just use the `beds` table though? 

(same applies to boats, bones, bucket, etc)

**Comment by est31** [comment]
@nerzhul yes that format is good.
@ShadowNinja what about git [subtrees](http://blogs.atlassian.com/2013/05/alternatives-to-git-submodule-git-subtree/)? I like submodules, but people have the concern, that when you download something as zip from github, they aren't bundled. As an example how it would look like, look at [my repo](https://github.com/est31/minetest_game/tree/test_intllib_subtree).

**Comment by Wuzzy2** [comment]
No, please do not merge this.

First: Intllib sucks too much. It does not support basic features one would expect from a decent translation system. For instance: Proper pluralization, like done in Gettext.
Second: This would only part of a certain subgame. But mod translation is such an essential feature it should **not** be done by subgames, but rather by the engine itself. Or by builtin. Otherwise, every mod has to duplicate all the work you have done this, which would lead to redundancy and inconsistency. The maintenance overhead would be a nightmare.
Third: The way inttllib works now, this can't be done without engine change. If you just use inttlib, this means all translations are server-side and all clients are forced to use the same language as the server uses.

Also: Please take into account the following discussions:
* https://forum.minetest.net/viewtopic.php?f=5&t=10031
* https://github.com/minetest/minetest/issues/2270

**Comment by est31** [comment]
Don't forget there is dual too in some languages, not just plural.
Wuzzy, the situation doesn't get better if we say "ah this is shit, close your pull request and go away". 
That doesn't encourage people.
Subgames already duplicate some work in other areas. Aren't tools basic too?
This is just a first step in a long process. Yes, I know an engine change is required. This PR however gives some first steps on the process:

1. people can start making translations right now
2. singleplayer is translated
3. most definitely a future engine change will use the groundwork done here to enable per client translations.

**Comment by Wuzzy2** [comment]
Sorry, my mistake. With “Pluralization support” I actually meant of course all kinds of grammatical numbers, and Gettext does this thing right.

Even if this PR will make a few things better in the short run, it will make things worse in the long run. As soon as Minetest has proper mods translation support, all this stuff needs to be reverted anyways.



Also: Are you aware of how messy intllib is to use? It is a pain in the ass to juggle around all those files, you have to **manually** update all those template files, I am not aware of any sort of automation here. I bet it is very easy here to break something by forgetting a string etc. I know this because I speak from experience.

In contrast, Gettext has, like, 1000s of tools to do all sorts of stuff with the files, and those tools are important. Intllib is just a quickly hacked solution.

But if you want intllib into Minetest so badly, then intllib should first have serious improvements. At the very least, it should have some support of grammatical number like in Gettext.
(On a sidenote: Why not just using Gettext right away? Minetest itself already uses Gettext, and I heard there are Lua bindings for Gettext. Sadly, I don't know about this binding and how to use it, but please keep it in mind.)

And another one already asked the important question of who is going to maintain this, once this got in. Your answer is …?

> Subgames already duplicate some work in other areas. 

That's a bad practice and should not be done. But sometimes it can't be avoided. In this case, it can very well avoided. As I said, if you want intllib so badly, then at least it should go into builtin, and not into Minetest Game.

> people can start making translations right now

True, but Minetest Game has not many strings anyways. The time overhead of this is really not worth mentioning.

> singleplayer is translated

True, but so are servers, forcibly translated into the language of the Minetest client / operating system. Have fun explaining this to server operators. This is not an issue which cannot be dealt with, I guess, but it is an issue, and should be fixed. Otherwise, this PR is un-mergable IMO.

> most definitely a future engine change will use the groundwork done here to enable per client translations.

Sorry, I don't see how.

> Wuzzy, the situation doesn't get better if we say "ah this is shit, close your pull request and go away".
> That doesn't encourage people, and we are throwing away precious contributions.

Development is not always unicorns and flowers. There are many good ideas around here, but also many bad ones. When I think I have spotted a bad idea, I will call it a bad idea. I did not say I do not appreciate that some work is done here, but I think that the work done here is leading into the wrong direction.

**Comment by est31** [comment]
Unmergeable is a very strong word. If you want to contribute, you can propose your own patch of how to do it. Just talking doesn't get things done.

**Comment by Adimgar** [comment]
There has been an attempt to do the same thing with gettext 
https://github.com/minetest/minetest_game/pull/470
If using gettext is a better option and more usefull in the long term, I'll try to convert this intllib work into gettext if you like and see what happens.
About forcing clients to use the game in a language, well, it's already that way with english, there are a lot of people that doesn't understand english and would like to play a game in their local language.
Inernational servers should definitively use english, but giving the option to offer servers in local languages is also important and not possible at this point.

**Comment by est31** [comment]
Yes, admittedly @Wuzzy2 has brought some very strong points for gettext. I'll have to find out what the problem is on android with gettext.

**Comment by paramat** [comment]
@Adimgar Rebase needed.

**Comment by paramat** [comment]
But before that there seems to be some discussion around this that needs to be resolved?

**Comment by Wuzzy2** [comment]
I am still advocating native mod translation support for Minetest instead of just hacking it into a single subgame.

**Comment by paramat** [comment]
I'm :-1:  on this PR, i agree with Wuzzy2, let's wait and do translation properly.

**Comment by sofar** [comment]
I'm less than enthusiastic on translations in general, and I'd really love to see translations be part of `minetest` and not `minetest_game` to start (and then proliferate from there). So no thumbs up from me on this, sorry.

**Comment by paramat** [comment]
@rubenwardy any opinion?

**Comment by paramat** [comment]
@Adimgar will close in a week if no reply.

**Comment by rubenwardy** [comment]
I only speak English and I am not especially knowledgeable about localisation. I think that it would be much better to add proper engine support for this, so different clients can have different languages.

Whilst this PR is still valid for singleplayer, and I also think that an engine solution would ideally also have intlib backwards compatibility support [1], I feel like for maintenance reasons it would be better to wait (a reasonable time) for engine support. If engine support for this will arrive around about the same time as client side modding _(ie: the heat death of the universe)_ then it would be better to have this than nothing.

Overall, I'm personally fairly neutral over this. I agree with Wuzzy on this. No thumbs up from me either. No thumbs down either, though.

------------------------

[1] Intlib backwards compatibility could be done by providing a dummy intlib mod which would read the locale files and create a `S()` function which would return whatever format the translations are transferred in, for example `English\0list,of,all,translated,string,versions`. This assumes that translations are transferred using the same format in node defs, formspecs and chat, as S() cannot tell where it is called from and where it will be used.

**Comment by Adimgar** [comment]
Hi all.

Thank you for your comments and the interest still shown in this Project.

As pointed out at the begining of the Project, it seems that using intlib for a "core" multilanguage support was not the best way to do it. I started a gettext implementation with no great success and had some lack of time to do a deeper research.

If you still see some future on the intlib translation as a starting point for a proper gettext implementation or something, i'll update all objets within this week for the rebase, but I don't have any problema if you decide to close this branch.

I still think that internationalization should be done and if someone comes out with a better way of doing it, i'll be pleased on helping with that.

**Comment by paramat** [comment]
Thanks for your work and reply, sorry this wasn't popular, closing due to lack of approval.

**Pull request closed without merging**